streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -1029,20 +1048,40 @@ int commit(final Collection<Task> tasksToCommit) {
             return -1;
         } else {
             int committed = 0;
-            final Map<TaskId, Map<TopicPartition, OffsetAndMetadata>> consumedOffsetsAndMetadataPerTask = new HashMap<>();
-            for (final Task task : tasksToCommit) {
+            final Map<Task, Map<TopicPartition, OffsetAndMetadata>> consumedOffsetsAndMetadataPerTask = new HashMap<>();
+            final Iterator<Task> it = tasksToCommit.iterator();
```

Comment:
I was initially worried about potential side-effects of removing from the input collection, but on second thought, maybe it's reasonable (considering the usage of this method) to assume that the input collection is always a copy already.

Still, it looks like we could actually simplify the implementation a little by _not_ reusing the input collection but instead only relying on the key set of `consumedOffsetsAndMetadataPerTask` as the list of "tasks to commit".

Here's what I'm thinking:
```java
int committed = 0;
            final Map<Task, Map<TopicPartition, OffsetAndMetadata>> consumedOffsetsAndMetadataPerTask = new HashMap<>();
            for (final Task task : tasksToCommit) {
                if (task.commitNeeded()) {
                    final Map<TopicPartition, OffsetAndMetadata> offsetAndMetadata = task.prepareCommit();
                    if (task.isActive()) {
                        consumedOffsetsAndMetadataPerTask.put(task, offsetAndMetadata);
                    }
                }
            }

            final Set<Task> uncommittedTasks = new HashSet<>();
            try {
                commitOffsetsOrTransaction(consumedOffsetsAndMetadataPerTask);
            } catch (final TaskTimeoutExceptions taskTimeoutExceptions) {
                final TimeoutException timeoutException = taskTimeoutExceptions.timeoutException();
                if (timeoutException != null) {
                    consumedOffsetsAndMetadataPerTask.keySet().forEach(t -> t.maybeInitTaskTimeoutOrThrow(time.milliseconds(), timeoutException));
                    uncommittedTasks.addAll(tasksToCommit);
                } else {
                    for (final Map.Entry<Task, TimeoutException> timeoutExceptions : taskTimeoutExceptions.exceptions().entrySet()) {
                        final Task task = timeoutExceptions.getKey();
                        task.maybeInitTaskTimeoutOrThrow(time.milliseconds(), timeoutExceptions.getValue());
                        uncommittedTasks.add(task);
                    }
                }
            }

            for (final Task task : tasksToCommit) {
                if (!uncommittedTasks.contains(task)) {
                    ++committed;
                    task.postCommit(false);
                    task.clearTaskTimeout();
                }
            }

            return committed;
```

It struck me as potentially incorrect to
* maybe init the timeout exception for standby tasks
* not clear the timeout for successful tasks when other tasks were timed out

But I'm also not sure whether these are big deals or not.

Reply:
Good catch!